src/core/mempool.rs

```diff
@@ -18,18 +63,411 @@ pub struct MempoolAdmitter {
+        
+        for cmd in MEMPOOL_SQL {
+            tx.execute(cmd, NO_PARAMS).map_err(db_error::SqliteError)?;
+        }
+
+        tx.commit().map_err(db_error::SqliteError)?;
+        Ok(())
+    }
+
+    /// Open the mempool db within the chainstate directory.
+    /// The chainstate must be instantiated already.
+    pub fn open(mainnet: bool, chain_id: u32, chainstate_path: &str) -> Result<MemPoolDB, db_error> {
+        match fs::metadata(chainstate_path) {
+            Ok(md) => {
+                if !md.is_dir() {
+                    return Err(db_error::NotFoundError);
+                }
+            }
+            Err(_e) => {
+                return Err(db_error::NotFoundError);
+            }
+        }
+
+        let chainstate = StacksChainState::open(mainnet, chain_id, chainstate_path)
+            .map_err(|e| db_error::Other(format!("Failed to open chainstate: {:?}", &e)))?;
+        
+        let mut path = PathBuf::from(chainstate.root_path.clone());
+
+        let admitter = MemPoolAdmitter::new(chainstate, BlockHeaderHash([0u8; 32]), BurnchainHeaderHash([0u8; 32]));
+
+        path.push("mempool.db");
+        let db_path = path.to_str().ok_or_else(|| db_error::ParseError)?.to_string();
+
+        let mut create_flag = false;
+        let open_flags =
+            if fs::metadata(&db_path).is_err() {
+                // need to create 
+                create_flag = true;
+                OpenFlags::SQLITE_OPEN_READ_WRITE | OpenFlags::SQLITE_OPEN_CREATE
+            }
+            else {
+                // can just open 
+                OpenFlags::SQLITE_OPEN_READ_WRITE
+            };
+
+        let mut conn = DBConn::open_with_flags(&db_path, open_flags).map_err(db_error::SqliteError)?;
+
+        if create_flag {
+            // instantiate!
+            MemPoolDB::instantiate_mempool_db(&mut conn)?;
+        }
+        
+        Ok(MemPoolDB {
+            db: conn,
+            path: db_path.to_string(),
+            admitter: admitter,
+        })
+    }
+
+    pub fn conn(&self) -> &DBConn {
+        &self.db
+    }
+
+    pub fn tx_begin<'a>(&'a mut self) -> Result<MemPoolTx<'a>, db_error> {
+        let tx = self.db.transaction().map_err(db_error::SqliteError)?;
+        Ok(MemPoolTx::new(tx, &mut self.admitter))
+    }
+
+    fn db_has_tx(conn: &DBConn, txid: &Txid) -> Result<bool, db_error> {
+        let has_tx_opt = conn.query_row("SELECT 1 FROM mempool WHERE txid = ?1", &[txid as &dyn ToSql],
+            |_row| {
+                Ok(true)
+            })
+            .optional()
+            .map_err(db_error::SqliteError)?;
+
+        match has_tx_opt {
+            Some(Ok(b)) => Ok(b),
+            Some(Err(e)) => Err(e),
+            None => Ok(false)
+        }
+    }
+
+    pub fn get_tx(conn: &DBConn, txid: &Txid) -> Result<Option<MemPoolTxInfo>, db_error> {
+        let tx_opt = conn.query_row("SELECT * FROM mempool WHERE txid = ?1", &[txid as &dyn ToSql],
```

Comment:
These functions should probably use util::db::query_row as well

Reply:
Fixed in 0cbbfb1